src/StreamJsonRpc/MessagePackFormatter.cs

```diff
@@ -1706,6 +1766,73 @@ public void Serialize(ref MessagePackWriter writer, Protocol.JsonRpcRequest valu
                 {
                     writer.WriteNil();
                 }
+
+                if (value.TraceParent?.Length > 0)
+                {
+                    TraceParentPropertyName.Write(ref writer);
+                    options.Resolver.GetFormatterWithVerify<TraceParent>().Serialize(ref writer, new TraceParent(value.TraceParent), options);
+
+                    if (value.TraceState?.Length > 0)
+                    {
+                        TraceStatePropertyName.Write(ref writer);
+                        WriteTraceState(ref writer, value.TraceState);
+                    }
+                }
+            }
+
+            private static void WriteTraceState(ref MessagePackWriter writer, string traceState)
+            {
+                ReadOnlySpan<char> traceStateChars = traceState.AsSpan();
+
+                // Count elements first so we can write the header.
+                int elementCount = 1;
+                int commaIndex;
+                while ((commaIndex = traceStateChars.IndexOf(',')) >= 0)
+                {
+                    elementCount++;
+                    traceStateChars = traceStateChars.Slice(commaIndex + 1);
+                }
+
+                // For every element, we have a key and value to record.
+                writer.WriteArrayHeader(elementCount * 2);
+
+                traceStateChars = traceState.AsSpan();
+                while ((commaIndex = traceStateChars.IndexOf(',')) >= 0)
+                {
+                    ReadOnlySpan<char> element = traceStateChars.Slice(0, commaIndex);
+                    int equalsIndex = element.IndexOf('=');
+                    ReadOnlySpan<char> key = element.Slice(0, equalsIndex);
+                    ReadOnlySpan<char> value = element.Slice(equalsIndex + 1);
+                    writer.Write(key);
+                    writer.Write(value);
+
+                    traceStateChars = traceStateChars.Slice(commaIndex + 1);
+                }
+            }
+
+            private static unsafe string ReadTraceState(ref MessagePackReader reader)
+            {
+                int elements = reader.ReadArrayHeader();
+                if (elements % 2 != 0)
+                {
+                    throw new NotSupportedException("Odd number of elements not expected.");
+                }
+
+                // With care, we could probably assemble this string with just two allocations (the string + a char[]).
+                var resultBuilder = new StringBuilder();
```

Comment:
I wonder if there is a cheap way to give the StringBuilder the know final size of the string so it never needs to resize it's internal buffer.

Reply:
There's a way, but it ain't cheap in terms of coding. per my code comment above, we could fully predict the length sufficient to allocate just `string` and one `char[]`. But the code is so complex that it seemed to add unnecessary risk. If/when we ever actually use `tracestate` (and I doubt we ever will) then we can certainly revisit this.